src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -126,96 +124,172 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
-                    OperationOutcomeConstants.IssueSeverity.Error,
-                    OperationOutcomeConstants.IssueType.Exception,
-                    ex.Message));
+                lock (jobLock)
+                {
+                    _reindexJobRecord.Error.Add(new OperationOutcomeIssue(
+                        OperationOutcomeConstants.IssueSeverity.Error,
+                        OperationOutcomeConstants.IssueType.Exception,
+                        ex.Message));
 
-                _reindexJobRecord.FailureCount++;
+                    _reindexJobRecord.FailureCount++;
 
-                _logger.LogError(ex, $"Encountered an unhandled exception. The job failure count increased to {_reindexJobRecord.FailureCount}.");
+                    _logger.LogError(ex, $"Encountered an unhandled exception. The job failure count increased to {_reindexJobRecord.FailureCount}.");
 
-                await UpdateJobAsync(cancellationToken);
+                    UpdateJobAsync(cancellationToken).Wait();
 
-                if (_reindexJobRecord.FailureCount >= _reindexJobConfiguration.ConsecutiveFailuresThreshold)
-                {
-                    await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
+                    if (_reindexJobRecord.FailureCount >= _reindexJobConfiguration.ConsecutiveFailuresThreshold)
+                    {
+                        CompleteJobAsync(OperationStatus.Failed, cancellationToken).Wait();
+                    }
                 }
             }
         }
 
-        private async Task ProcessQueryAsync(ReindexJobQueryStatus query, CancellationToken cancellationToken)
+        private async Task<ReindexJobQueryStatus> ProcessQueryAsync(ReindexJobQueryStatus query, object jobLock, CancellationToken cancellationToken)
         {
             try
             {
-                query.Status = OperationStatus.Running;
-                query.LastModified = DateTimeOffset.UtcNow;
+                SearchResult results;
 
-                // Query first batch of resources
-                var results = await ExecuteReindexQueryAsync(query, false, cancellationToken);
-
-                // if continuation token then update next query
-                if (!string.IsNullOrEmpty(results.ContinuationToken))
+                lock (jobLock)
                 {
-                    var nextQuery = new ReindexJobQueryStatus(results.ContinuationToken);
-                    nextQuery.LastModified = DateTimeOffset.UtcNow;
-                    nextQuery.Status = OperationStatus.Queued;
-                    _reindexJobRecord.QueryList.Add(nextQuery);
-                }
+                    query.Status = OperationStatus.Running;
+                    query.LastModified = DateTimeOffset.UtcNow;
 
-                await UpdateJobAsync(cancellationToken);
+                    // Query first batch of resources
+                    results = ExecuteReindexQueryAsync(query, false, cancellationToken).Result;
```

Comment:
I don't think we should be combining synchronous blocks with async/await pattern. It defeats the purpose of using async/await

Reply:
Changed to use semaphore instead of lock(object) to properly enable async/await pattern.